### PR TITLE
fea/ir support fuse, based on graph pattern detection helper

### PR DESCRIPTION
**Pull request by Superjomn**
- unified graph trivial algorithm support
  - unlike the existing `TopologySortOperations`, the `GraphTraits` is an iterator, and it can visit `Var` as well as `Op`
  - the graph pattern is defined with both the `Op` and `Var`s
  - one can use the pattern detector to find some `Var`s and `Op`s that match some conditions.

- A pattern is combined by 
  - some pattern nodes with conditions called `PDNode`
  - some edges that define the connections between the pattern nodes
  - so one can use this helper to find any
     - `ir::Node`s matches some conditions
     - IR subgraphs matches some conditions

It can be used in kernel fusion scenios with some specific rules, for example:
  - CONV + BATCHNORM => BATCHNORM

Some other conditions , search for the `CPU` -> `GPU` copying, for example:
  - Ops in CPU place -> var -> Ops in GPU place

and so on.

### CODE REVIEW

File: paddle/fluid/framework/ir/graph_traits.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+/*
+ * GraphTraits contains some graph traversal algorithms.
+ *
+ * Usage:
+ *
+ */
+struct GraphTraits {
+  static iterator_range<NodesDFSIterator> DFS(const Graph &g) {
+    auto start_points = InferenceStartPoints(g);
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph_traits.h]: Does this need to be inference specific?

File: paddle/fluid/framework/ir/graph_traits.h [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    NodesDFSIterator x(start_points);
+    return iterator_range<NodesDFSIterator>(NodesDFSIterator(start_points),
+                                            NodesDFSIterator());
+  }
+
+ private:
+  static std::vector<Node *> InferenceStartPoints(const Graph &g) {
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph_traits.h]: Is this inference specific?

File: paddle/fluid/framework/ir/graph_traits.cc [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+NodesDFSIterator &NodesDFSIterator::operator++() {
+  if (stack_.empty()) return *this;
+  visited_.insert(stack_.top());
+  auto *cur = stack_.top();
+  stack_.pop();
+  for (auto *x : cur->outputs) {
+    if (!visited_.count(x)) {
+      stack_.push(x);
+      visited_.insert(x);
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph_traits.cc]: Why is visited inserted here and above twice?

File: paddle/fluid/framework/ir/graph_pattern_detecter.cc [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  if (graph.Nodes().empty()) return false;
+
+  for (auto& node : GraphTraits::DFS(graph)) {
+    for (const auto& pdnode : pattern_.nodes()) {
+      if (pdnode->Tell(&node)) {
+        pdnodes2nodes_[pdnode.get()].insert(&node);
+      }
+    }
+  }
+  return !pdnodes2nodes_.empty();
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph_pattern_detecter.cc:62]: return value seems not used?

File: paddle/fluid/framework/ir/graph_pattern_detecter.cc [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+  }
+  return false;
+}
+
+std::vector<GraphPatternDetecter::subgraph_t>
+GraphPatternDetecter::DetectPatterns() {
+  // Init empty subgraphs.
+  std::vector<GraphPatternDetecter::subgraph_t> result;
+  std::vector<HitGroup> init_groups;
+  auto* first_pnode = pattern_.edges().front().first;
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph_pattern_detecter.cc]: CHECK pattern_ not empty?

File: paddle/fluid/framework/ir/graph_traits.cc [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+NodesDFSIterator::NodesDFSIterator(const NodesDFSIterator &other)
+    : stack_(other.stack_), visited_(other.visited_) {}
+
+Node &NodesDFSIterator::operator*() {
+  PADDLE_ENFORCE(!stack_.empty());
+  return *stack_.top();
+}
+
+NodesDFSIterator &NodesDFSIterator::operator++() {
+  if (stack_.empty()) return *this;
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph_traits.cc]: should throw exception?

File: paddle/fluid/framework/ir/graph_traits.cc [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  stack_.pop();
+  for (auto *x : cur->outputs) {
+    if (!visited_.count(x)) {
+      stack_.push(x);
+      visited_.insert(x);
+    }
+  }
+  return *this;
+}
+bool NodesDFSIterator::operator==(const NodesDFSIterator &other) {
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph_traits.cc:52]: Is this method needed?

File: paddle/fluid/framework/ir/graph_traits.h [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+  iterator_range(const IteratorT &begin, const IteratorT &end)
+      : begin_(begin), end_(end) {}
+
+  const IteratorT &begin() const { return begin_; }
+  const IteratorT &end() const { return end_; }
+};
+
+// DFS iterator on nodes.
+struct NodesDFSIterator
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph_traits.h:39]: should be documented that it's from input to output.

File: paddle/fluid/framework/ir/graph_pattern_detecter.h [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+  std::vector<std::unique_ptr<PDNode>> nodes_;
+  std::vector<edge_t> edges_;
+};
+
+/*
+ * GraphPatternDetecter helps to detect the specific patterns in the graph.
+ * Input a pattern, output a list of the matched subgraphs/nodes.
+ * This helper can be used to support fuse(conv+batchnorm => batchnorm e.g.).
+ */
+class GraphPatternDetecter {
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph_pattern_detecter.h]: Briefly document the algorithm used?

File: paddle/fluid/framework/ir/graph_pattern_detecter.h [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+ public:
+  using hit_rcd_t =
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph_pattern_detecter.h]: what does rcd mean?

File: paddle/fluid/framework/ir/graph_pattern_detecter.h [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+ private:
+  // Mark the nodes that fits the pattern.
+  bool MarkPDNodesInGraph(const ir::Graph& graph);
+
+  // Detect all the pattern and output the hit records.
+  std::vector<subgraph_t> DetectPatterns();
+
+  // Remove duplicate patterns
+  void UniquePatterns(std::vector<subgraph_t>* subgraphs);
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph_pattern_detecter.h]: What's the strategy when there are overlapped match? Should be documented?

File: paddle/fluid/framework/ir/graph_pattern_detecter.cc [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+    HitGroup group;
+    group.roles[first_pnode] = node;
+    init_groups.emplace_back(group);
+  }
+
+  int step = 0;
+  std::array<std::vector<HitGroup>, 2> bi_records;
+  bi_records[0] = std::move(init_groups);
+
+  for (const auto& edge : pattern_.edges()) {
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph_pattern_detecter.cc]: document this code block?

File: paddle/fluid/framework/ir/graph_pattern_detecter.cc [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+      if (pdnode->Tell(&node)) {
+        pdnodes2nodes_[pdnode.get()].insert(&node);
+      }
+    }
+  }
+  return !pdnodes2nodes_.empty();
+}
+
+struct HitGroup {
+  std::unordered_map<PDNode*, Node*> roles;
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph_pattern_detecter.cc:66]: wht call it roles?

File: paddle/fluid/framework/ir/graph_pattern_detecter.cc (hunk 4 above)
Comment (Superjomn) [reply paddle/fluid/framework/ir/graph_pattern_detecter.cc:62]: done; it should be used in operator() to check whether there are any nodes those are marked.

File: paddle/fluid/framework/ir/graph_pattern_detecter.cc (hunk 13 above)
Comment (Superjomn) [reply paddle/fluid/framework/ir/graph_pattern_detecter.cc:66]: Node -> PDNode

a Node will be mapped to a PDNode which represents something, this phase likes assign a role to a Node.

File: paddle/fluid/framework/ir/graph_pattern_detecter.cc [hunk 14]
```diff
@@ -151,6 +158,29 @@ void GraphPatternDetecter::UniquePatterns(
...
+    std::vector<subgraph_t>* subgraphs) {
+  std::vector<subgraph_t> result;
+  std::unordered_set<Node*> node_set;
+
+  for (const auto& subgraph : *subgraphs) {
+    bool valid = true;
+    for (auto& item : subgraph) {
+      if (node_set.count(item.second)) {
+        break;
+        valid = false;
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph_pattern_detecter.cc]: bug?

File: paddle/fluid/framework/ir/graph_traits.cc (hunk 7 above)
Comment (Superjomn) [reply paddle/fluid/framework/ir/graph_traits.cc:52]: used when check `it == end()`

**Review by panyx0718 — DISMISSED** [review]

File: paddle/fluid/framework/ir/CMakeLists.txt [hunk 15]
```diff
@@ -3,7 +3,10 @@ cc_library(graph SRCS graph.cc DEPS node)
 cc_library(graph_helper SRCS graph_helper.cc DEPS graph)
 cc_library(pass SRCS pass.cc DEPS graph node graph_helper)
 cc_library(graph_viz_pass SRCS graph_viz_pass.cc DEPS graph pass graph_helper)
+cc_library(graph_traits SRCS graph_traits.cc DEPS graph)
+cc_library(graph_pattern_detecter SRCS graph_pattern_detecter.cc DEPS graph graph_helper graph_traits)
 
 cc_test(pass_test SRCS pass_test.cc DEPS graph pass graph_helper)
 cc_test(graph_test SRCS graph_test.cc DEPS graph graph_helper op_registry)
 cc_test(graph_helper_test SRCS graph_helper_test.cc DEPS graph graph_helper op_registry)
+cc_test(test_graph_pattern_detecter SRCS graph_pattern_detecter_tester.cc DEPS graph_pattern_detecter)
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/CMakeLists.txt:12]: change the name to test.cc and graph_xxx_test

**Review by panyx0718 — APPROVED** [review]

**Pull request merged**